b2g/components/RemoteControlService.jsm

```diff
@@ -0,0 +1,395 @@
+            let socket;
+            for (let i = 100; i; i--) {
+              let temp = Cc["@mozilla.org/network/tls-server-socket;1"].createInstance(Ci.nsITLSServerSocket);
+              temp.init(self._port, false, MAX_CLIENT_CONNECTIONS);
+              temp.serverCert = cert;
+
+              let allowed = ios.allowPort(temp.port, "tls");
+              if (!allowed) {
+                DEBUG && debug("Warning: obtained TLSServerSocket listens on a blocked port: " + temp.port);
+              }
+
+              if (!allowed && self._port == -1) {
+                DEBUG && debug("Throw away TLSServerSocket with bad port.");
+                temp.close();
+                continue;
+              }
+
+              socket = temp;
+              break;
+            }
+
+            if (!socket) {
+              throw new Error("No socket server available. Are there no available ports?");
+            }
+
+            DEBUG && debug("Listen on port " + socket.port + ", " + MAX_CLIENT_CONNECTIONS + " pending connections");
+
+            socket.serverCert = cert;
+            socket.setSessionCache(false);
+            socket.setSessionTickets(false);
+            socket.setRequestClientCertificate(Ci.nsITLSServerSocket.REQUEST_NEVER);
+
+            socket.asyncListen(self);
+            self._port = socket.port;
+            self._socket = socket;
+          } catch (e) {
+            DEBUG && debug("Could not start server on port " + self._port + ": " + e);
+            aReject("Start TLSSocketServer fail");
+          }
+
+          // Register mDNS remote control service with this._port
+          if (("@mozilla.org/toolkit/components/mdnsresponder/dns-sd;1" in Cc)) {
+            let serviceInfo = Cc["@mozilla.org/toolkit/components/mdnsresponder/dns-info;1"].createInstance(Ci.nsIDNSServiceInfo);
+            serviceInfo.serviceType = "_remotecontrol._tcp";
+            serviceInfo.serviceName = Services.prefs.getCharPref("dom.presentation.device.name");
+            serviceInfo.port = self._port;
+
+            let mdns = Cc["@mozilla.org/toolkit/components/mdnsresponder/dns-sd;1"].getService(Ci.nsIDNSServiceDiscovery);
+            self._mDNSCancelableHandler = mdns.registerService(serviceInfo, null);
+          }
+
+          aResolve();
+          self._serverStatus = SERVER_STATUS.STARTED;
+        }
+      }
+    });
+  },
+
+  // Notifies this server that the given connection has been closed.
+  _connectionClosed: function(connection) {
+    NS_ASSERT(connection.number in this._connections,
+              "closing a connection " + this + " that we never added to the " +
+              "set of open connections?");
+    NS_ASSERT(this._connections.get(connection.number) === connection,
+              "connection number mismatch?  " +
+              this._connections.get(connection.number));
+    delete this._connections.get(connection.number);
```

Comment:
NS_ASSERT is defined in debug.js. I think we can just print which connection is closed to:
_connectionClosed: function(connection) {
    DEBUG && debug("Close connection " + connection.number);
    this._connections.delete(connection.number);
  },